### DISCUSSION TITLE
Please help with .S and .inc files 😢

### DISCUSSION THREAD

**Discussion by aabmets** [python-cffi/cffi#60] [Q&A] [answered]
Greetings!

I'm trying to use CFFI to compile [these C source files](https://github.com/PQClean/PQClean/tree/master/crypto_sign/dilithium5/avx2) and I'm just at a loss on how to feed the `.S` and `.inc` files 
into the `ffi.set_source()` method so they would be correctly compiled with the rest of the `.h` and `.c` files. 
[This](https://github.com/aabmets/quantcrypt/blob/main/scripts/compile.py) is the compile script I'm trying to make it work with and [here](https://github.com/aabmets/quantcrypt/blob/3b349db4f9804ddb68f03e5f3c793f5b891e635e/scripts/compile.py#L204) is the `ffi.set_source()` line. 
Any help would be greatly appreciated, as I really don't know how to make this work. Thanks in advance!

**Comment by aabmets** [comment, marked as answer]
Okay, I solved it myself, literally 16 minutes after posting this 😄 and 
ChatGPT-4 was _absolutely useless_ in solving this problem. I happened on 
the solution by trying out random solutions in an absolute desperation.

The solution is to pass the `.S` files as a `list[str]` of absolute paths into the `extra_objects` kwarg 
and the absolute path of the directory of the C source code into the `extra_link_args` kwarg 
as `list[str]` in this format: `"-I/path/to/source/directory"` of the `ffi.set_source()` method. 

The compiled code works perfectly.

**Comment by arigo** [comment]
Great that you figured it out!  More generally, you can use any argument in `ffi.set_source()`, it ends up passed to setuptools.  Your solution certainly works; maybe you could also have found in the setuptools documentation if they have some explicit support for .S files.
